### DISCUSSION TITLE
Feature Request: Detection of NaN values

### DISCUSSION THREAD

**Discussion by dominikbell** [pyccel/pyccel#1408] [Ideas]
I want to detect `NaN` values in a file that is to be "pyccelized" (translated to fortran and compiled). As far as I know and understood, there is currently no way of doing that. When trying to import the function `isnan`, or the types `NaN` or `nan` from `numpy`, the compilation command throws the fatal error
```
Function 'isnan' from module 'numpy' is not currently supported by pyccel (from numpy import isnan)
```
when e.g. trying to import `isnan`.

I am not sure if this counts as a "non-trivial feature" but on the issue creation of a "Feature Request" it told me to link to a discussion which should take place here. I have not found a discussion, PR, or issue concerning this feature.

**Comment by EmilyBourne** [comment]
I think `NaN` detection can be non-trivial in Fortran.
In C it is simple. There is a `isnan` function in `math.h` (https://en.cppreference.com/w/c/numeric/math/isnan).
However in Fortran it can be compiler-dependent. GNU provides an isnan function (https://gcc.gnu.org/onlinedocs/gcc-4.8.5/gfortran/ISNAN.html) as does Intel (https://www.intel.com/content/www/us/en/docs/fortran-compiler/developer-guide-reference/2023-0/isnan.html). I have also seen people use `a /= a` (returns True if `a` is `NaN` but is often optimised out by the compiler).

That said I have just discovered the function `ieee_is_nan` from the `ieee_arithmetic` intrinsic module which was introduced in Fortran 2003. In my opinion this module could be used to implement the NumPy functions:
- `isnan`
- `isfinite`
- `isinf`

@yguclu What is your opinion on this solution?

**Reply by yguclu** [reply]
@EmilyBourne I am absolutely in favor of using an intrinsic module for this. How much do you know about the compilers' support for `ieee_arithmetic`?

**Reply by EmilyBourne** [reply]
According to this: https://fortranwiki.org/fortran/show/Fortran+2003+status it is supported by all major compilers

**Reply by EmilyBourne** [reply]
![image](https://github.com/pyccel/pyccel/assets/25363473/3b8b54c7-088e-4021-b28b-95dbc1404e68)
This is the earliest versions I can see in the source documents. These seem early enough for `ieee_arithmetic` to not be a problem

**Reply by yguclu** [reply]
Excellent! I think we can go forward and create dedicated issues.
